### PR TITLE
CTE: support explain CTE plan

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

more tests will be added after CTE executor get merged

https://github.com/pingcap/tidb/issues/24111

some examples:

```sql
mysql> explain with cte (a) as (select 1) select * from cte;
+-------------------------------+---------+------+---------------+--------------------+
| id                            | estRows | task | access object | operator info      |
+-------------------------------+---------+------+---------------+--------------------+
| CTEFullScan_9                 | 1.00    | root | CTE:cte       | data:CTE_0         |
| CTE_0                         | 1.00    | root |               | None Recursive CTE |
| └─Projection_7(Seed Part)     | 1.00    | root |               | 1->Column#2        |
|   └─TableDual_8               | 1.00    | root |               | rows:1             |
+-------------------------------+---------+------+---------------+--------------------+
4 rows in set (0.00 sec)

mysql> explain with recursive cte1 (c1) as (with recursive cte2 (b) as (select 1 union select b + 1 from cte2 where b < 100) select * from cte2 union select c1 + 1 from cte1 where c1 < 10) select * from cte1 where c1 < 10;
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| id                                  | estRows | task | access object | operator info                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| Selection_32                        | 0.80    | root |               | lt(Column#9, 10)                                    |
| └─CTEFullScan_43                    | 1.00    | root | CTE:cte1      | data:CTE_0                                          |
| CTE_0                               | 1.00    | root |               | Recursive CTE                                       |
| ├─Projection_33(Seed Part)          | 1.00    | root |               | Column#7                                            |
| │ └─CTEFullScan_39                  | 1.00    | root | CTE:cte2      | data:CTE_2                                          |
| └─Projection_40(Recursive Part)     | 0.80    | root |               | cast(plus(Column#9, 1), bigint(1) BINARY)->Column#9 |
|   └─Selection_41                    | 0.80    | root |               | lt(Column#9, 10)                                    |
|     └─CTETable_42                   | 1.00    | root |               | Scan on CTE_0                                       |
| CTE_2                               | 1.00    | root |               | Recursive CTE                                       |
| ├─Projection_34(Seed Part)          | 1.00    | root |               | 1->Column#7                                         |
| │ └─TableDual_35                    | 1.00    | root |               | rows:1                                              |
| └─Projection_36(Recursive Part)     | 0.80    | root |               | cast(plus(Column#7, 1), bigint(1) BINARY)->Column#7 |
|   └─Selection_37                    | 0.80    | root |               | lt(Column#7, 100)                                   |
|     └─CTETable_38                   | 1.00    | root |               | Scan on CTE_2                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
14 rows in set (0.00 sec)

mysql> explain with recursive cte (a) as (select 1 union select a + 1 from cte where a < 10) select * from cte t1, cte t2;
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| id                                  | estRows | task | access object | operator info                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| HashJoin_16                         | 1.00    | root |               | CARTESIAN inner join                                |
| ├─CTEFullScan_24(Build)             | 1.00    | root | CTE:t2        | data:CTE_0                                          |
| └─CTEFullScan_23(Probe)             | 1.00    | root | CTE:t1        | data:CTE_0                                          |
| CTE_0                               | 1.00    | root |               | Recursive CTE                                       |
| ├─Projection_18(Seed Part)          | 1.00    | root |               | 1->Column#3                                         |
| │ └─TableDual_19                    | 1.00    | root |               | rows:1                                              |
| └─Projection_20(Recursive Part)     | 0.80    | root |               | cast(plus(Column#3, 1), bigint(1) BINARY)->Column#3 |
|   └─Selection_21                    | 0.80    | root |               | lt(Column#3, 10)                                    |
|     └─CTETable_22                   | 1.00    | root |               | Scan on CTE_0                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
9 rows in set (0.01 sec)

mysql> explain with recursive cte1 (c1) as (with recursive cte2 (a) as (select 1 union select a + 1 from cte2 where a < 5) select * from cte2 where a < 1 union select c1 + 1 from cte1 where c1 < 10) select * from cte1 where c1 < 2;
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| id                                  | estRows | task | access object | operator info                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
| Selection_34                        | 0.80    | root |               | lt(Column#9, 2)                                     |
| └─CTEFullScan_46                    | 1.00    | root | CTE:cte1      | data:CTE_0                                          |
| CTE_0                               | 1.00    | root |               | Recursive CTE                                       |
| ├─Projection_35(Seed Part)          | 0.80    | root |               | Column#7                                            |
| │ └─Selection_36                    | 0.80    | root |               | lt(Column#7, 1)                                     |
| │   └─CTEFullScan_42                | 1.00    | root | CTE:cte2      | data:CTE_2                                          |
| └─Projection_43(Recursive Part)     | 0.80    | root |               | cast(plus(Column#9, 1), bigint(1) BINARY)->Column#9 |
|   └─Selection_44                    | 0.80    | root |               | lt(Column#9, 10)                                    |
|     └─CTETable_45                   | 1.00    | root |               | Scan on CTE_0                                       |
| CTE_2                               | 1.00    | root |               | Recursive CTE                                       |
| ├─Projection_37(Seed Part)          | 1.00    | root |               | 1->Column#7                                         |
| │ └─TableDual_38                    | 1.00    | root |               | rows:1                                              |
| └─Projection_39(Recursive Part)     | 0.80    | root |               | cast(plus(Column#7, 1), bigint(1) BINARY)->Column#7 |
|   └─Selection_40                    | 0.80    | root |               | lt(Column#7, 5)                                     |
|     └─CTETable_41                   | 1.00    | root |               | Scan on CTE_2                                       |
+-------------------------------------+---------+------+---------------+-----------------------------------------------------+
15 rows in set (0.00 sec)
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- support explain CTE plan

### CODE REVIEW

File: planner/core/physical_plans.go [hunk 1]
```diff
@@ -595,6 +598,7 @@ type PhysicalProjection struct {
 	Exprs                []expression.Expression
 	CalculateNoDelay     bool
 	AvoidColumnEvaluator bool
+	AvoidEliminateForCTE bool
```
Comment (wjhuang2016) [inline planner/core/physical_plans.go:601]: No need for it.

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -1031,6 +1037,27 @@ func (e *Explain) RenderResult() error {
 	return nil
 }
 
+func (e *Explain) explainPlanInRowFormatCTE() (err error) {
+	explainedCTEPlan := make(map[int]struct{})
+	for i := 0; i < len(e.ctes); i++ {
+		x := (*CTEDefinition)(e.ctes[i])
+		if _, ok := explainedCTEPlan[x.CTE.IDForStorage]; ok {
```
Comment (wjhuang2016) [inline planner/core/common_plans.go:1044]: Add a comment for it.

File: planner/core/encode.go [hunk 3]
```diff
@@ -59,9 +61,34 @@ func (pn *planEncoder) encodePlanTree(p Plan) string {
...
+			continue
+		}
+		taskTypeInfo := plancodec.EncodeTaskType(true, kv.TiKV)
+		actRows, analyzeInfo, memoryInfo, diskInfo := getRuntimeInfo(x.SCtx(), x, nil)
+		rowCount := 0.0
+		if statsInfo := x.statsInfo(); statsInfo != nil {
+			rowCount = x.statsInfo().RowCount
+		}
+		plancodec.EncodePlanNode(0, x.CTE.IDForStorage, plancodec.TypeCTEDefinition, rowCount, taskTypeInfo, x.ExplainInfo(), actRows, analyzeInfo, memoryInfo, diskInfo, &pn.buf)
+		if x.SeedPlan != nil {
```
Comment (wjhuang2016) [inline planner/core/encode.go:82]: Can it be nil?

File: planner/core/common_plans.go (hunk 2 above)
Comment (xiongjiwei) [reply planner/core/common_plans.go:1045]: updated

File: planner/core/encode.go (hunk 3 above)
Comment (xiongjiwei) [reply planner/core/encode.go]: never be nil 😅

**Review by winoros — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**